Draft: v2.1.0

hwiedPro opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
### Python Module:
- ci:
      - Support for Azure Pipelines
      - Split ci pushes to Failure, Success and Production Branch, so that all pipeline pushes can be reviewed before production
      - Extended Test-protocols in PRs
      - Improve configurability of model_def by hierarchical defaults
      - autogenerated files can be marked as such
      - Results from feature/* branches will be pushed to the corresponding feature/* branch in the target repo, for testing purposes
- scripts:
      - phobos convert: Fix for Scenes/Worlds, Args for mesh type and export config
- core/io:
      - Refactor utils
      - Adaption in defaults.json
      - Several bugfixes in mesh handling
      - Bugfixes for poses
      - Configurability for exports has been extended
      - Enhanced submodel export
      - Bugfixes for import/export
- smurf:
      - Include list of submodels
      - A mesh collision can now be annotated with primitive collisions, during export the user can decide what to export
- general:
      - Several error messages have been improved
      - Minor codestyle improvements
      - Updated ROS templates for ROS2
    
### Blender:
- Only minor bugfixes
- Most bugfixes in Python Module apply here as well
    
### ToDo:
  - [ ] Documentation has to be updated accordingly

## Related Issues/PRs
#363 #352 #357 #354

## Motivation and Context
See Changelog above

## How Has This Been Tested?
- Scripts and Pipeline have been tested in an integrated environment
- [ ] @AlpenAalAlex Would you be so kind to test this at your end?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, and the tests you ran to -->
<!--- see how your change affects other areas of the code, etc. -->

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.

hwiedPro committed:
Update defaults.json

Add topic entry
https://github.com/dfki-ric/phobos/issues/330

hwied committed:
Fix #339
https://github.com/dfki-ric/phobos/issues/339

hwied committed:
Fix #331
https://github.com/dfki-ric/phobos/issues/331

AlpenAalAlex committed:
Fix #338 (2)

AlpenAalAlex committed:
Fix link pose not exported in sdf #338 (1) #339

Exporting pose relative to root link because gazebo 9 ignores relative_to attribute

AlpenAalAlex committed:
Fix link pose not exported in sdf #338 (1) #339

Exporting pose relative to root link because gazebo 9 ignores relative_to attribute

hwiedPro committed:
Merge branch 'master' into develop

hwiedPro committed:
Fix install_requirements.py

hwiedPro committed:
Add frame handling

hwiedPro committed:
Fix to_hey_color()

hwiedPro committed:
Catch git_error when getting maintainer, author, url

hwiedPro committed:
Get log level from env

hwiedPro committed:
Add -m option to run pipeline for specific definition file

hwiedPro committed:
Add possibility to define default_export_config in pipeline.yml

hwiedPro committed:
Add verbose option to execute_shell_command

hwiedPro committed:
Make to_pretty_xml_string more robust

hwiedPro committed:
Adapt append_string to logger usage

hwiedPro committed:
Fix setup_git script

hwiedPro committed:
Fix Git-LFS Readme integration

hwiedPro committed:
Fix several path/file handling issues

hwiedPro committed:
Fix model_testing

hwiedPro committed:
Ensure nothing is changed on export()

hwiedPro committed:
Refactored version of Mathias commit: https://github.com/dfki-ric/phobos/commit/3c833ce46a1b8ff7c78babc9d02e693c8f54a28f
Co-authored-by: mathias.trampler@dfki.de

AlpenAalAlex committed:
Fix urdf export

AlpenAalAlex committed:
Enable Add Motor operator for batch selection

Previously a link had to be the active object for the operator to be enabled. Now a link has to be in the selection and the operator will be applied to all links in the selection

AlpenAalAlex committed:
Apply Add Motor operator only to joints

Add feedback

AlpenAalAlex committed:
Update message boxes

AlpenAalAlex committed:
Redraws the sidebar when changing phobostype

AlpenAalAlex committed:
Fix tooltip

AlpenAalAlex committed:
Adds search to phobos panel

AlpenAalAlex committed:
Add gearbox joints

AlpenAalAlex committed:
Sort parameters

AlpenAalAlex committed:
Merge branch 'issue324-Undefined-key-joint/type-when-exporting' into issue195-Support-for-multiple-degrees-of-freedom-joints

AlpenAalAlex committed:
Update editing.py

AlpenAalAlex committed:
Add ball joint object

AlpenAalAlex committed:
Add ball joint

AlpenAalAlex committed:
Set unused joint parameters to None

AlpenAalAlex committed:
Add universal joint

AlpenAalAlex committed:
Add screw joint draft

AlpenAalAlex committed:
Add screw joint

AlpenAalAlex committed:
Export screw joint thread pitch

AlpenAalAlex committed:
Remove screwdrivers before defining joints

AlpenAalAlex committed:
Screw joint tested in gazebo

AlpenAalAlex committed:
Add screw joint object

AlpenAalAlex committed:
Unset unused gearbox and screw variables

AlpenAalAlex committed:
Add gearbox joint object

yarohenning committed:
Merge remote-tracking branch 'origin/master' into develop

yarohenning committed:
[BREAKING CHANGES] Squashed refactoring of python module

Python Module:
--------------
ci:
  - Support for Azure Pipelines
  - Split ci pushes to Failure, Success and Production Branch, so that all pipeline pushes can be reviewed before production
  - Extended Test-protocols in PRs
  - Improve configurability of model_def by hierarchical defaults
  - autogenerated files can be marked as such
  - Results from feature/* branches will be pushed to the corresponding feature/* branch in the target repo, for testing purposes
scripts:
  - phobos convert: Fix for Scenes/Worlds, Args for mesh type and export config
core/io:
  - Refactor utils
  - Adaption in defaults.json
  - Several bugfixes in mesh handling
  - Bugfixes for poses
  - Configurability for exports has been extended
  - Enhanced submodel export
  - Bugfixes for import/export
smurf:
  - Include list of submodels
  - A mesh collision can now be annotated with primitive collisions, during export the user can decide what to export
general:
  - Several error messages have been improved
  - Minor codestyle improvements
  - Updated ROS templates for ROS2

Blender:
--------
  - Only minor bugfixes
  - Most bugfixes in Python Module apply here as well

ToDo:
-----
  - Documentation has to be updated accordingly

yarohenning committed:
Merge remote-tracking branch 'origin/master' into pre_v2.1.0

hwiedPro committed:
Fix #363

hwiedPro committed:
Bugfix for 646ceaf

AlpenAalAlex committed:
Merge branch 'issue195-Support-for-multiple-degrees-of-freedom-joints' of https://github.com/AlpenAalAlex/phobos into issue195-Support-for-multiple-degrees-of-freedom-joints

AlpenAalAlex commented on phobos/utils/misc.py:
```diff
@@ -142,17 +142,23 @@ def is_binary_file(filepath):
     if dry_run:
         return "", ""
-    proc = subprocess.Popen([cmd], stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True, cwd=cwd)
+    proc = subprocess.Popen([cmd], stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True, cwd=cwd, executable="/bin/bash")
```
This does not work on Windows. For Windows the code can look like this:
proc = subprocess.Popen(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, shell=True, cwd=cwd)
which changes
[cmd] -> cmd
executable can be omitted or usually C:\\Windows\system32\cmd.exe

hwiedPro replied:
Ok thanks for the test. I'll put the executable as argument as it has a very specific use case.

AlpenAalAlex commented on phobos/ci/base_model.py:
```diff
@@ -811,11 +811,10 @@ def deploy(self, mesh_commit, failed_model=False, uses_lfs=False):
+                readme_content = open(readme_path, "r").read()
+                if "Git LFS" not in readme_content:
+                    with open(readme_path, "a") as readme:
+                        readme.write(open(resources.get_resources_path("GitLFS_README.md.in")).read())
```
Faulty indentation

AlpenAalAlex reviewed (dismissed):
Left comments on:
[phobos/ci/base_model.py](https://github.com/dfki-ric/phobos/pull/364/commits/6492934febb1115c1314ea740861fbf7d12f4bdf#diff-77065fd661bd806462286202defb07c88ea26d27d13e0503f91be26cad81b4fa):948
[phobos/utils/misc.py](https://github.com/dfki-ric/phobos/pull/364/commits/9db3b92cfd9e3bad96766ac793a8d5a93df0c8f4#diff-ea4e02022d3afe07d40322dc6023219b711f4b0962d4b3a2243b48dda01d4b6f):291

Everything else looks good 👍

AlpenAalAlex committed:
Add velocity2 and effort2

AlpenAalAlex committed:
Add spring and damping to universal joints

hwiedPro committed:
Fix execute_shell_command for Windows

hwiedPro committed:
Fix faulty indentation

hwiedPro commented:
Thanks for your review, I fixed the issues you have found.

hwiedPro committed:
Bump version to 2.1.0

AlpenAalAlex committed:
Implements TODO update rigid body constraints

AlpenAalAlex committed:
Remove axis from ball joint

AlpenAalAlex committed:
Add universal joint object

AlpenAalAlex committed:
Show orthogonal axis for universal joints

AlpenAalAlex committed:
Comments and a fix

AlpenAalAlex committed:
Merge branch 'Issue351-Create-Inertials-fails-if-not-all-vertices-are-selected' into issue195-Support-for-multiple-degrees-of-freedom-joints

AlpenAalAlex committed:
Fix export fails in edit mode

AlpenAalAlex committed:
Fixes

AlpenAalAlex committed:
Fix error if child and parent are parented in opposite order

AlpenAalAlex committed:
Fix #363 - 2

AlpenAalAlex committed:
Fix #366

AlpenAalAlex committed:
Add TODO: gearbox drivers

AlpenAalAlex committed:
Fix tooltip

AlpenAalAlex committed:
Fix Add Sensor tooltip

AlpenAalAlex committed:
Fix Remove Motor info box

AlpenAalAlex committed:
Update sidebar when setting geometry type

AlpenAalAlex committed:
Fix Select by Name operator

by skipping unselectable objects that are not visible (joints)

AlpenAalAlex committed:
Fix Select Submechanism operator

AlpenAalAlex committed:
Merge angle parameters

AlpenAalAlex committed:
Convert angles when toggling radian checkbox

AlpenAalAlex committed:
Fix submechanism import

AlpenAalAlex committed:
Fix motor import

AlpenAalAlex committed:
Adds valid default export settings

AlpenAalAlex committed:
Update phobosgui.py

AlpenAalAlex committed:
Hide delete button for Add Sensor op

AlpenAalAlex committed:
Fix #370

m0rsch1 committed:
phobos/core/multiple.py: Check if file is given before calling any other func

Signed-off-by: Moritz Schilling <moritz.schilling@dfki.de>

m0rsch1 committed:
phobob/core/multiple.py: file property was asking for a wrong attribute of self.model

Signed-off-by: Moritz Schilling <moritz.schilling@dfki.de>

m0rsch1 committed:
phobos/core/multiple.py: export entities for SMURFA

Signed-off-by: Moritz Schilling <moritz.schilling@dfki.de>

m0rsch1 committed:
phobos/io/representation: Fixed position setter which was failing when given a dict with x,y,z keys and values

Signed-off-by: Moritz Schilling <moritz.schilling@dfki.de>

m0rsch1 committed:
phobos/scripts/convert.py: Fixed bad argument to robot.export()

Signed-off-by: Moritz Schilling <moritz.schilling@dfki.de>

m0rsch1 commented:
@hwiedPro @AlpenAalAlex i have some huge fixes regarding to SMURFA export, etc. Can i directly push into this branch (if given access to) or do i have to fork?

m0rsch1 committed:
phobos/core/multiple.py: Retrieve name from file

Signed-off-by: Moritz Schilling <moritz.schilling@dfki.de>

m0rsch1 committed:
phobos/core/multiple.py: Added FIXME statement to assemble function

This function does not properly assemble a consistent URDF/SMURF from an
Arrangement!

Signed-off-by: Moritz Schilling <moritz.schilling@dfki.de>

m0rsch1 committed:
phobos/core/multiple.py: The toplvl arrangement name has to be used for proper export

Also, interface renaming has to be done according to root_entity only

This runs through but my example produces a weird URDF (which makes MARS crash)

Signed-off-by: Moritz Schilling <moritz.schilling@dfki.de>

m0rsch1 committed:
Adjusted phobos/scripts/convert.py to properly support SMURFA to directory conversion
(which is a SMURFA to SMURF conversion actually)

Signed-off-by: Moritz Schilling <moritz.schilling@dfki.de>

m0rsch1 committed:
Show resulting origin info

Signed-off-by: Moritz Schilling <moritz.schilling@dfki.de>

m0rsch1 committed:
Looks like exchange_root() is working (again?) :D

To be tested

Signed-off-by: Moritz Schilling <moritz.schilling@dfki.de>

m0rsch1 committed:
Clean up

Signed-off-by: Moritz Schilling <moritz.schilling@dfki.de>

m0rsch1 committed:
Clean up II

Signed-off-by: Moritz Schilling <moritz.schilling@dfki.de>

m0rsch1 committed:
When inverting a joint in exchange_root() the axis also has to be transformed

Signed-off-by: Moritz Schilling <moritz.schilling@dfki.de>

m0rsch1 committed:
Fixed output format detection if path contains a single "."

Signed-off-by: Moritz Schilling <moritz.schilling@dfki.de>

Commit:
Fixed double not in phobos/scripts/convert.py

AlpenAalAlex committed:
Merge pull request #373 from m0rsch1/fix_smurfa_processsing

Fixes to SMURFA io and conversion

AlpenAalAlex committed:
Fix mesh collision created at world origin #375

AlpenAalAlex committed:
Fix material creation - Material output missing #375

AlpenAalAlex committed:
Fix export with mirrored objects #377

AlpenAalAlex committed:
Merge branch 'pre_v2.1.0' into issue338-Exporting-scaled-meshes

AlpenAalAlex committed:
Rename parameter to origin_root

AlpenAalAlex committed:
Merge branch 'pre_v2.1.0' into issue195-Support-for-multiple-degrees-of-freedom-joints

AlpenAalAlex committed:
Merge branch 'pre_v2.1.0' into aaldev

AlpenAalAlex committed:
Merge pull request #348 from AlpenAalAlex/issue212-Add-Motor-function

Fix Add Motor Operator

AlpenAalAlex committed:
Merge pull request #344 from AlpenAalAlex/issue338-Exporting-scaled-meshes

Fix exporting scaled meshes - add link pose to sdf export

AlpenAalAlex committed:
Rename parameter

AlpenAalAlex committed:
Merge pull request #371 from AlpenAalAlex/aaldev

Develop

AlpenAalAlex committed:
Merge branch 'pre_v2.1.0' into issue195-Support-for-multiple-degrees-of-freedom-joints

AlpenAalAlex committed:
Update xml_formats.json

AlpenAalAlex committed:
Merge pull request #368 from AlpenAalAlex/issue195-Support-for-multiple-degrees-of-freedom-joints

Support for multiple degrees of freedom joints

AlpenAalAlex committed:
Add glb mesh type #369

AlpenAalAlex committed:
Merge pull request #382 from AlpenAalAlex/issue377-URDF-Export-Error-and-Collision-Object-Error-with-Left-Handed-Rotation-Matrix

Fix export with mirrored objects

AlpenAalAlex committed:
Merge pull request #381 from AlpenAalAlex/issue375-Problems-with-create-collision-object-and-create-inertial

Fix mesh collision created at world origin / material creation

AlpenAalAlex committed:
Support Blender 4.1 export API

AlpenAalAlex committed:
Support Blender 4.1 import

AlpenAalAlex committed:
Merge pull request #384 from AlpenAalAlex/issue369-glb-import-and-export

Add glb mesh type

AlpenAalAlex committed:
Merge pull request #385 from AlpenAalAlex/issue361-Blender-4-2

Support Blender 4.2

AlpenAalAlex committed:
Add experimental note to bobj

AlpenAalAlex committed:
Set geometry type when setting phobos type with op

AlpenAalAlex committed:
Merge pull request #389 from AlpenAalAlex/issue388-AttributeError-The-geometry-of-object-Cube-has-not-yet-been-defined

Set geometry type when setting phobos type with operator

AlpenAalAlex committed:
Merge pull request #386 from AlpenAalAlex/issue370-Error-in-install_requirements-py

Fix installation of requirements

PimpMyPizza committed:
Ignore vscode workspace files

PimpMyPizza committed:
Added a version to compare versions and avoid duplicated code

PimpMyPizza committed:
Fix blender crashing issue due to infinite recursion when clicking on the button "Create Inertials"

PimpMyPizza committed:
Fix error when exporting urdf with bobj files when _mesh_information is None

PimpMyPizza committed:
Export .mtl file with .obj if textures are checked in the UI. Use blender comparison function to avoid duplicate code

PimpMyPizza committed:
Use relative paths for .mtl export is the user set the right setting for it

PimpMyPizza committed:
Adapt axis forward and UP to match the expected values for the obj export

PimpMyPizza committed:
Simpler way to convert axis_forward and axis_up

malter committed:
Merge pull request #392 from PimpMyPizza/pre_v2.1.0

Some fixes to pre_v2.1.0

m0rsch1 committed:
When importing blenderVersionIsAtLeast or other bpy dependent stuff
guard it properly

Signed-off-by: Moritz Schilling <moritz.schilling@dfki.de>

malter committed:
Merge pull request #403 from m0rsch1/pre_v2.1.0

When importing blenderVersionIsAtLeast or other bpy dependent stuff guard it properly
